feat: livestream (cast) system

dudantas opened the pull request:
This pull request introduces a new Livestream system to the game, allowing players to broadcast their gameplay and receive experience bonuses while streaming. It includes configuration options, database schema updates, server-side logic, and player commands for managing livestreams. Additionally, there are minor bugfixes and refactoring in unrelated files.

**Livestream System Implementation**

* Added new configuration options in `config.lua.dist` and code support for livestream experience multipliers, viewer limits, and minimum caster level. [[1]](diffhunk://#diff-41ebed0867fb6bc8676568dd635813d639f92062d0101ccd9cc64658d35bb594R463-R472) [[2]](diffhunk://#diff-50caaf697dd4fd21c3eec347ee00594a435825407f627abd3b2bee6084fcaba5R124-R128) [[3]](diffhunk://#diff-355085a4323ac4d4defe3385378854a991c8d45678628d2dc710b9e1b974a002R20-R27) [[4]](diffhunk://#diff-355085a4323ac4d4defe3385378854a991c8d45678628d2dc710b9e1b974a002R272-R276)
* Created new database table `active_livestream_casters` and updated migration/versioning logic (`schema.sql`, `46.lua`, `47.lua`). [[1]](diffhunk://#diff-939acd018b931292330e519becd8d003ed0dd20c5e42de62ec3d37553698739aR256-R263) [[2]](diffhunk://#diff-b6f7aa51b9db5a3d4984cffaea1b9b6e71375c60643ee43f7be89b5dbc72ff8dL2-R12) [[3]](diffhunk://#diff-9586f5f07cde8ae5cb337ec7ce2670f73729fd642f274d3c062087ed3805a6ccR1-R3) [[4]](diffhunk://#diff-939acd018b931292330e519becd8d003ed0dd20c5e42de62ec3d37553698739aL10-R10)
* Implemented a comprehensive Lua script (`livestream_system.lua`) providing player commands to manage livestreams, including starting/stopping streams, viewer management, password protection, description setting, and experience bonus handling.
* Integrated server-side support for the Livestream system, including new source files and configuration loading.

**Minor Bugfixes and Refactoring**

* Fixed typos in monster attack definitions (`lenght` → `length`) in monster scripts. [[1]](diffhunk://#diff-8060ad60d343279176b7ef5519823ceb635a3d49fdebb950a43f7a20949dfd24L66-R66) [[2]](diffhunk://#diff-dcc8d2abf64954475341813be0199a2a7761b4a4699e154168ecbf28dc652402L110-R110)
* Updated chat channel message sending to include an additional parameter.

Extensive testing and reviews are required to ensure the correct functioning of the system. 

In my tests everything is working.

New login for 13x (won't work in Tibia 13 without this login)
[login.zip](https://github.com/user-attachments/files/17107110/login.zip)


Works on
- [x] Tibia 11 outdated
- [x] Otclient
- [x] Tibia 13+

dudantas committed:
feat: cast system

phacUFPE commented on src/creatures/players/cast/cast_viewer.cpp:
```diff
@@ -0,0 +1,1550 @@
+	}
+	broadcast_message << second << " seconds.";
+	return broadcast_message.str();
+}
```
I think that it could be simplier showing the hours and minutes even if they are lesser than 1;
```suggestion
std::string CastViewer::getCastBroadcastTimeString() const {
	int64_t seconds = getCastBroadcastTime() / 1000;
	uint16_t hour = floor(seconds / 60 / 60 % 24);
	uint16_t minute = floor(seconds / 60 % 60);
	uint16_t second = floor(seconds % 60);

	return fmt::format("{} hours, {} minutes and {} seconds", hour, minute, second);
}
```

phacUFPE commented on src/creatures/players/cast/cast_viewer.cpp:
```diff
@@ -0,0 +1,1550 @@
+			sendChannelMessage("", "New record: " + std::to_string(m_castLiveRecord) + " people are watching your livestream now.", TALKTYPE_CHANNEL_O, CHANNEL_CAST);
+		}
+	}
+}
```
What about this:
```suggestion
void CastViewer::addViewer(ProtocolGame_ptr client, bool spy) {
	if (m_viewers.find(client) != m_viewers.end()) {
		return;
	}

	auto viewerId = m_viewers.size() + 1;
	std::string guestString = fmt::format("Guest-{}", viewerId);

	m_viewers[client] = std::make_pair(guestString, m_id);

	if (!spy) {
		sendChannelMessage("", fmt::format("{} has entered the cast.", guestString), TALKTYPE_CHANNEL_O, CHANNEL_CAST);

		if (m_viewers.size() > m_castLiveRecord) {
			m_castLiveRecord = m_viewers.size();
			sendChannelMessage("", fmt::format("New record: {} people are watching your livestream now.", std::to_string(m_castLiveRecord), TALKTYPE_CHANNEL_O, CHANNEL_CAST);
		}
	}
}
```
?

phacUFPE commented on src/creatures/players/cast/cast_viewer.cpp:
```diff
@@ -0,0 +1,1550 @@
+	}
+
+	if (!spy) {
+		sendChannelMessage("", it->second.first + " has left the cast.", TALKTYPE_CHANNEL_O, CHANNEL_CAST);
```
I don't think that here will be really necessary, but if you want to make and keep the usage of fmt::format pattern.
```suggestion
		sendChannelMessage("", fmt::format("{} has left the cast.", it->second.first), TALKTYPE_CHANNEL_O, CHANNEL_CAST);
```

phacUFPE commented on src/creatures/players/cast/cast_viewer.cpp:
```diff
@@ -0,0 +1,1550 @@
+	} else if (client->m_castCount++ >= 3) {
+		std::stringstream messageViewer;
+		messageViewer << "Please wait a " << ((client->m_castCooldownTime + 5000 - now) / 1000) + 1 << " seconds to send another message.";
+		client->sendTextMessage(TextMessage(MESSAGE_STATUS, messageViewer.str()));
```
Same about the keeping the pattern:
```suggestion
		std::string messageViewer = fmt::format("Please wait a {} seconds to send another message.", ((client->m_castCooldownTime + 5000 - now) / 1000) + 1;
		client->sendTextMessage(TextMessage(MESSAGE_STATUS, messageViewer));
```

mattyx14 commented:
i writte on spanish ...
esto esta de lujo gran acierto 

great dan

phacUFPE commented on src/creatures/players/cast/cast_viewer.cpp:
```diff
@@ -0,0 +1,1550 @@
+			}
+
+			messageViewer << ".";
+			client->sendTextMessage(TextMessage(MESSAGE_STATUS, messageViewer.str()));
```
I don't know if it is the best way to do this, but with a vector of viewers name we could use `fmt::join` and will increase the readability.
```suggestion
			auto viewersNames = std::views::transform(m_viewers, [](const auto &pair) {
                                return pair.second.first;
                        });
			std::string messageViewer = fmt::format("{} spectator{}, {}.", m_viewers.size(), m_viewers.size() > 1 ? "s" : "", fmt::join(viewersNames.begin(), viewersNames.end(), ", "));

			client->sendTextMessage(TextMessage(MESSAGE_STATUS, messageViewer));
```

phacUFPE commented on src/creatures/players/cast/cast_viewer.cpp:
```diff
@@ -0,0 +1,1550 @@
+
+						if (!found) {
+							if (isCastChannel) {
+								sendChannelMessage("", sit->second.first + " was renamed for " + CommandParam[1] + ".", TALKTYPE_CHANNEL_O, CHANNEL_CAST);
```
Same about the pattern:
```suggestion
								sendChannelMessage("", fmt::format("{} was renamed to {}.", sit->second.first, CommandParam[1]), TALKTYPE_CHANNEL_O, CHANNEL_CAST);
```

phacUFPE reviewed (commented):
Adding some suggestions.

phacUFPE commented on src/server/network/protocol/protocolgame.cpp:
```diff
@@ -9031,3 +9095,137 @@ void ProtocolGame::sendDisableLoginMusic() {
+	if (player->client->isCastBroadcasting()) {
+		std::stringstream WelcomeMSG;
+		WelcomeMSG << player->getName() << " is broadcasting for " << player->client->getViewers().size() << " people.\nLivestream time: " << player->client->getCastBroadcastTimeString();
+		sendTextMessage(TextMessage(MESSAGE_LOOK, WelcomeMSG.str().c_str()));
```
```suggestion
                std::string welcomeMessage = fmt::format("{} is boradcasting for {} people.\nLivestream time: {}", player->getName(), player->client->getViewers().size(), player->client->getCastBroadcastTimeString());
		sendTextMessage(TextMessage(MESSAGE_LOOK, welcomeMessage));
```

phacUFPE commented on src/server/network/protocol/protocollogin.cpp:
```diff
@@ -179,3 +191,56 @@ void ProtocolLogin::onRecvFirstMessage(NetworkMessage &msg) {
+	output->addByte(0x14);
+	std::ostringstream ss;
+	ss << normal_random(1, 100) << "\nWelcome to Cast System!";
+	output->addString(ss.str(), "ProtocolLogin::getCastViewersList - ss.str()");
```
This would have the same effect?
```suggestion
	output->addString(fmt::format("{}\nWelcome to Cast System!", normal_random(1, 100)), "ProtocolLogin::getCastViewersList - fmt::format('{}\nWelcome to Cast System!', normal_random(1, 100))");
```

phacUFPE commented on src/server/network/protocol/protocollogin.cpp:
```diff
@@ -179,3 +191,56 @@ void ProtocolLogin::onRecvFirstMessage(NetworkMessage &msg) {
+
+	// Add session key
+	output->addByte(0x28);
+	output->addString("@cast\n" + password, "ProtocolLogin::getCharacterList - accountDescriptor + password");
```
```suggestion
	output->addString(fmt::format("@cast\n{}", password), "ProtocolLogin::getCharacterList - accountDescriptor + password");
```

phacUFPE commented on src/server/network/protocol/protocollogin.cpp:
```diff
@@ -179,3 +191,56 @@ void ProtocolLogin::onRecvFirstMessage(NetworkMessage &msg) {
+
+	for (const auto &player : players) {
+		output->addByte(0);
+		output->addString(player->getName(), "ProtocolLogin::getCastViewersList - ss.str()");
```
```suggestion
		output->addString(player->getName(), "ProtocolLogin::getCastViewersList - player->getName()");
```

phacUFPE reviewed (commented):
Finishing the suggestions and questions,

elsongabriel commented on data/scripts/talkactions/player/cast_system.lua:
```diff
@@ -0,0 +1,347 @@
+
+castOnStartup:register()
+
+local gainExperience = EventCallback()
```
```suggestion
local gainExperience = EventCallback("PlayerOnGainExperienceCastSystem")
```

dudantas replied:
The EventCallback modification is still in pull request

cys9211 commented:
> FAILED: src/CMakeFiles/canary_lib.dir/Unity/unity_2_cxx.cxx.o 
/usr/bin/c++ -DCURL_STATICLIB -DNDEBUG -DSPDLOG_COMPILED_LIB -DSPDLOG_FMT_EXTERNAL -I/home/Pulpit/canary-dudantas-feat-casting-system/src -I/home/Pulpit/canary-dudantas-feat-casting-system/build/linux-release/vcpkg_installed/x64-linux/include/luajit-2.1 -I/home/Pulpit/canary-dudantas-feat-casting-system/build/linux-release/src/protobuf -isystem /home/Pulpit/canary-dudantas-feat-casting-system/build/linux-release/vcpkg_installed/x64-linux/include/mysql -isystem /home/Pulpit/canary-dudantas-feat-casting-system/build/linux-release/vcpkg_installed/x64-linux/include -flto=auto -flto=auto -O2 -g -DNDEBUG -std=gnu++20 -flto=auto -fno-fat-lto-objects -fPIC -Wno-unused-parameter -Wno-sign-compare -Wno-switch -Wno-implicit-fallthrough -Wno-extra -Wno-deprecated-declarations -fopenmp -MD -MT src/CMakeFiles/canary_lib.dir/Unity/unity_2_cxx.cxx.o -MF src/CMakeFiles/canary_lib.dir/Unity/unity_2_cxx.cxx.o.d -o src/CMakeFiles/canary_lib.dir/Unity/unity_2_cxx.cxx.o -c /home/Pulpit/canary-dudantas-feat-casting-system/build/linux-release/src/CMakeFiles/canary_lib.dir/Unity/unity_2_cxx.cxx
In file included from /home/Pulpit/canary-dudantas-feat-casting-system/build/linux-release/src/CMakeFiles/canary_lib.dir/Unity/unity_2_cxx.cxx:7:
/home/Pulpit/canary-dudantas-feat-casting-system/src/creatures/players/cast/cast_viewer.cpp: In member function ‘void CastViewer::addViewer(ProtocolGame_ptr, bool)’:
/home/Pulpit/canary-dudantas-feat-casting-system/src/creatures/players/cast/cast_viewer.cpp:1416:186: error: expected ‘)’ before ‘;’ token
 1416 |                         sendChannelMessage("", fmt::format("New record: {} people are watching your livestream now.", std::to_string(m_castLiveRecord), TALKTYPE_CHANNEL_O, CHANNEL_CAST);
      |                                           ~                                                                                                                                              ^
      |                                                                                                                                                                                          )
/home/Pulpit/canary-dudantas-feat-casting-system/src/creatures/players/cast/cast_viewer.cpp: In member function ‘void CastViewer::handle(ProtocolGame_ptr, const string&, uint16_t)’:
/home/Pulpit/canary-dudantas-feat-casting-system/src/creatures/players/cast/cast_viewer.cpp:1453:164: error: expected ‘)’ before ‘;’ token
 1453 |                 std::string messageViewer = fmt::format("Please wait a {} seconds to send another message.", ((client->m_castCooldownTime + 5000 - now) / 1000) + 1;
      |                                                        ~                                                                                                           ^
      |                                                                                                                                                                    )
[24/31] Building CXX object src/CMakeF...canary_lib.dir/Unity/unity_4_cxx.cxx.o
ninja: build stopped: subcommand failed.


Problem on compile.

elsongabriel commented on src/creatures/players/cast/cast_viewer.cpp:
```diff
@@ -1459,9 +1459,8 @@ void CastViewer::handle(ProtocolGame_ptr client, const std::string &text, uint16
-		std::stringstream messageViewer;
-		messageViewer << "Please wait a " << ((client->m_castCooldownTime + 5000 - now) / 1000) + 1 << " seconds to send another message.";
-		client->sendTextMessage(TextMessage(MESSAGE_STATUS, messageViewer.str()));
+		std::string messageViewer = fmt::format("Please wait a {} seconds to send another message.", ((client->m_castCooldownTime + 5000 - now) / 1000) + 1;
```
```suggestion
    std::string messageViewer = fmt::format("Please wait a {} seconds to send another message.", ((client->m_castCooldownTime + 5000 - now) / 1000) + 1);
```

elsongabriel commented:
Something I notice:
| Problem / Print |
|--------|
| When the streamer opens the gamestore window, the spectators see it too. |
| ![image](https://github.com/opentibiabr/canary/assets/7812282/56c65075-8bc9-4fc5-88e2-aad1bca9eaec) |

| Problem / Print |
|--------|
| Spectator opening menus like cyclopedia, bestiary, highscore,  social, compendium, forge, etc.. (but, not load the info) |
| ![image](https://github.com/opentibiabr/canary/assets/7812282/ceacac7c-79da-4eb2-a41d-9448c778f2c6) |

| Problem / Print |
|--------|
| Is it possible to change the Tibia title, adding Tibia Cast suffix? like `Tibia - Playername (cast)` | 
| ![image](https://github.com/opentibiabr/canary/assets/7812282/fef00e77-194b-45e7-8f88-05f02e9750e7) |

Chapeleiiro commented:
Well, I am having a problem with the /effect command. When I use the command /effect 262, it does not give the correct effect, and it shows a different effect. Also, if I apply any new effect, it does not work.

phacUFPE commented on src/creatures/players/cast/cast_viewer.hpp:
```diff
@@ -0,0 +1,261 @@
+	void sendDistanceShoot(const Position &from, const Position &to, unsigned char type) const;
+	void sendCreatePrivateChannel(uint16_t channelId, const std::string &channelName);
+	void sendIcons(uint32_t icons) const;
+	void sendMagicEffect(const Position &pos, uint8_t type) const;
```
It should be 16 bits.

```suggestion
	void sendMagicEffect(const Position &pos, uint16_t type) const;
```

phacUFPE commented on src/creatures/players/cast/cast_viewer.cpp:
```diff
@@ -0,0 +1,1533 @@
+	}
+}
+
+void CastViewer::sendMagicEffect(const Position &pos, uint8_t type) const {
```
It should be 16 bits.
```suggestion
void CastViewer::sendMagicEffect(const Position &pos, uint16_t type) const {
```

phacUFPE reviewed (changes requested):
These changes are extremely needed because are causing bit overflow in the magic effect types.

OpenTibiaMX2 commented:
Can someone provide a login.php for gesior

OpenTibiaMX2 commented:
Can someone provide a login.php for gesior

jbracovich commented:
The Cast System is having issues with protocol 1100 which prevents any account from logging in because when trying to do so, it recognizes it as protocol 1332 instead of recognizing it as protocol 1100.

![image](https://github.com/opentibiabr/canary/assets/165123587/c87a7b96-8bcf-40a4-a1fc-988036c89361)

XmasGame commented:
![image](https://github.com/opentibiabr/canary/assets/50735276/d52d73ea-b6df-4446-8c56-3190e5158d82)

thats appaers.. why? (lastest canary-main)

XmasGame commented:
> The Cast System is having issues with protocol 1100 which prevents any account from logging in because when trying to do so, it recognizes it as protocol 1332 instead of recognizing it as protocol 1100.
> 
> ![image](https://private-user-images.githubusercontent.com/165123587/336390135-c87a7b96-8bcf-40a4-a1fc-988036c89361.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc1NDQ1ODEsIm5iZiI6MTcxNzU0NDI4MSwicGF0aCI6Ii8xNjUxMjM1ODcvMzM2MzkwMTM1LWM4N2E3Yjk2LThiY2YtNDBhNC1hMWZjLTk4ODAzNmM4OTM2MS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjA0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYwNFQyMzM4MDFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xMGU5YWYyZmVmMmRjYWY1NDkyYzE0NWU5YjU1NGVhMjY5MjNhYzk5NzAyNDcxYTQ0YTMzNWQwNDA3MmExYWVmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.7UvGfEHBTaaxgJhuHb09vXYNt1s6HMfORVrYpsk_VqM)

same problem...

XmasGame commented:
![image](https://github.com/opentibiabr/canary/assets/50735276/0e716084-b76f-4e4e-bdab-e37fb107d336)

its working cast <- only
when you want to loggin to play normally on tibia 11, doesn't work

Chapeleiiro commented:
I have two active pull requests for Soul War and Cast System, and I'm not sure if this is causing interference. Specifically, with the file cast_system.lua, the file prevents players from gaining XP, and elemental weapons do not deal elemental damage.

Other problems:
The player who is viewing the cast can open the following tabs: Skills, Battle List, Party List, VIP, Unjustified Points, Prey, Exaltation Forge, and Spell List, showing some information with the player's balance and others.

dudantas commented:
> I have two active pull requests for Soul War and Cast System, and I'm not sure if this is causing interference. Specifically, with the file cast_system.lua, the file prevents players from gaining XP, and elemental weapons do not deal elemental damage.
> 
> Other problems: The player who is viewing the cast can open the following tabs: Skills, Battle List, Party List, VIP, Unjustified Points, Prey, Exaltation Forge, and Spell List, showing some information with the player's balance and others.

I'm busy at the moment, but when I can I'll fix it, if you find more bugs please report them

jhogberg commented on src/server/network/protocol/protocolgame.cpp:
```diff
@@ -9031,3 +9095,136 @@ void ProtocolGame::sendDisableLoginMusic() {
+	acceptPackets = true;
+
+	if (player->client->isCastBroadcasting()) {
+		std::string welcomeMessage = fmt::format("{} is boradcasting for {} people.\nLivestream time: {}", player->getName(), player->client->getViewers().size(), player->client->getCastBroadcastTimeString());
```
Typo:
```suggestion
		std::string welcomeMessage = fmt::format("{} is broadcasting for {} people.\nLivestream time: {}", player->getName(), player->client->getViewers().size(), player->client->getCastBroadcastTimeString());
```

jhogberg commented on src/server/network/protocol/protocolgame.cpp:
```diff
@@ -9031,3 +9095,136 @@ void ProtocolGame::sendDisableLoginMusic() {
+			sendCreatureSay(player, TALKTYPE_SAY, description);
+		}
+
+		sendChannel(CHANNEL_CAST, "Tibia Cast", nullptr, nullptr);
```
The word "livestream" is used in many places, perhaps the channel should also be called that?

jhogberg commented on config.lua.dist:
```diff
@@ -452,6 +452,11 @@ combatChainSkillFormulaAxe = 0.9
 
+-- Cast system
+-- NOTE: set experience to 0 to deactivate the bonus
+castExperienceMultiplier = 15;
```
This name is inconsistent with the actual usage of `exp = exp * (1 + (experienceMultiplier / 100))`, which implies that this is a percentage bonus rather than a multiplier. Why not make this a multiplier where `1.0` means no bonus? It would also simplify some code later on.

dudantas replied:
Changed here: ae07440df 
Thanks for the review.

jhogberg commented on src/creatures/players/cast/cast_viewer.cpp:
```diff
@@ -0,0 +1,1533 @@
+	std::string t = _password;
+	trimString(t);
+
+	return t == m_castPassword;
```
Nitpick: this leaks the password through timing. Is there any rate limiting before this?

jhogberg commented on src/server/network/protocol/protocolgame.cpp:
```diff
@@ -811,6 +832,12 @@ void ProtocolGame::onRecvFirstMessage(NetworkMessage &msg) {
 		return;
 	}
 
+	// Cast system login (show casting players)
```
Is there any reason to handle viewer logins before checking (global) IP bans?

jhogberg commented on data/scripts/talkactions/player/cast_system.lua:
```diff
@@ -0,0 +1,347 @@
+	elseif table.contains({ "kick", "remove" }, split[1]) then
+		if data.broadcast then
+			if split[2] then
+				if split[2] ~= "all" then
```
What if a griefer names themselves `all`? :sweat_smile:

dudantas replied:
From my tests there were no problems. Do you have any other suggestions of test?

jhogberg replied:
If you have 3 viewers and one of them has named themselves `all`, what happens to the other 2 viewers when you `!kick all`?

jhogberg commented on data/scripts/talkactions/player/cast_system.lua:
```diff
@@ -0,0 +1,347 @@
+				if split[2]:len() > 0 and split[2]:len() <= 50 then
+					player:kv():scoped("cast-system"):set("description", split[2])
+				else
+					player:sendTextMessage(MESSAGE_STATUS, "Your description max lenght 50 characters.")
```
Typo:
```suggestion
					player:sendTextMessage(MESSAGE_STATUS, "Your description max length 50 characters.")
```

jhogberg commented on data/scripts/talkactions/player/cast_system.lua:
```diff
@@ -0,0 +1,347 @@
+	elseif table.contains({ "desc", "description" }, split[1]) then
+		if split[2] then
+			-- print(split[2])
+			if table.contains({ "remove", "delete" }, split[2]) then
```
Why not let an empty description mean its removal?

jhogberg commented on src/creatures/players/cast/cast_viewer.cpp:
```diff
@@ -0,0 +1,1533 @@
+	if (m_owner) {
+		m_owner->sendAddContainerItem(cid, slot, item);
+
+		for (const auto &it : m_viewers) {
```
Sending synchronously like this gives the player (`m_owner`) a latency hit, which may be light at 50 viewers but sets an upper bound on how many viewers a stream can have.

Have you considered sending the (serialized) data to an entity on another thread that, in turn, sends the data to the clients? Tibiacast worked this way and had a constant latency hit regardless of how many viewers a stream had (we sent it to our server instead of another thread, but the principle is the same).

dudantas replied:
Implemented here: b09fce73d, please if you can review and give me feedback.

jhogberg replied:
It's hard to say, does the dispatcher provide strong ordering guarantees under `TaskGroup::GenericParallel`? I would imagine tile operations (add, remove, update, etc) causing client issues if they happened to execute out of turn.

dudantas replied:
> It's hard to say, does the dispatcher provide strong ordering guarantees under `TaskGroup::GenericParallel`? I would imagine tile operations (add, remove, update, etc) causing client issues if they happened to execute out of turn.

Thank you for the feedback! I understand the concerns regarding potential ordering issues with TaskGroup::GenericParallel, especially with tile operations. I haven't fully analyzed the implications yet, but I will take a closer look into how the dispatcher manages the operation sequence to ensure it won’t cause issues for clients. Once I have a better understanding, I'll provide a more definitive response.

#Edit 
Thanks, I did some testing and it really needs to be the Serial to work properly: d4559d0b4

mehah replied:
@dudantas 
```c++
	// Make a copy of viewers in case they can be modified while being processed in another thread.
	threadPool.detach_task([viewers = m_viewers, func]() {
		for (const auto &[client, _] : viewers) {
			func(client);
		}
	});
```

jhogberg commented on src/creatures/players/cast/cast_viewer.cpp:
```diff
@@ -0,0 +1,1533 @@
+			it.first->sendCreatureSkull(creature);
+		}
+	}
+}
```
I can't help but notice a pattern in this code base where shared pointers are copied for functions that don't take ownership of the pointed object. `sendCreatureSkull` reads from the creature object but does not take ownership of it, so why not pass the shared pointer by const reference?

Here, the reference count on `creature` gets bounced once for the call to this function, once for sending to the player, and once for each viewer. It's also done again inside `ProtocolGame::sendAddCreature` itself when calling `Player::getSkullClient`. That's a lot of expensive atomic increment/decrement operations that could be skipped.

(Changing this in the protocol-related parts probably won't make much of a difference, but this pattern looks pervasive and may have a more significant impact in performance-sensitive parts.)

dudantas replied:
Hello @jhogberg, thank you very much for your detailed and insightful feedback!

You are absolutely right in pointing out that frequently passing std::shared_ptr by value, especially in functions that do not require ownership of the object, is not ideal due to the costs associated with atomic operations on the reference counter. Your suggestion to pass std::shared_ptr by const reference in cases where the function does not take ownership of the object is an excellent practice that we will definitely adopt.

I will review parts of the code where this practice can be applied to enhance performance and efficiency. Thanks again for bringing this to our attention!

beats-dh replied:
@jhogberg I'm doing this, you can test and follow along in the following pr: https://github.com/opentibiabr/canary/pull/2863/files

jhogberg replied:
Great :)

If I'm reading the code right there's another measure you can take for better performance, which is a lot more work but I think your current architecture could be adapted to it: [epoch-based reclamation](https://www.cl.cam.ac.uk/techreports/UCAM-CL-TR-579.pdf) (section 5.2.3). If you're the practical sort you might prefer to read https://github.com/erlang/otp/blob/master/erts/emulator/internal_doc/ThreadProgress.md and the accompanying implementation in https://github.com/erlang/otp/blob/master/erts/emulator/beam/erl_thr_progress.c

It gives you multi-threaded deletion safety without having to deal with reference counting, "free" memory barriers, and also all the nice parts of having a garbage collector with regards to lock-free algorithms (e.g. the [ABA problem](https://en.wikipedia.org/wiki/ABA_problem) becomes moot in scenarios like the compare-exchange loops in the push/pop operations in #2825) but without the issues that come with a GC.

Since most operations seem to go through `Dispatcher` I think that's a natural place to drive this stuff. It does require a change in how you think about most things, but opens the door for much better multi-core performance.

jhogberg commented on src/creatures/players/cast/cast_viewer.cpp:
```diff
@@ -0,0 +1,1533 @@
+			sendChannelMessage(sit->second.first, text, TALKTYPE_CHANNEL_Y, CHANNEL_CAST);
+		}
+	} else {
+		client->sendTextMessage(TextMessage(MESSAGE_FAILURE, "You are mutated."));
```
Typo:
```suggestion
		client->sendTextMessage(TextMessage(MESSAGE_FAILURE, "You are muted."));
```

jhogberg commented on src/creatures/players/cast/cast_viewer.cpp:
```diff
@@ -0,0 +1,1533 @@
+	if (client->m_castCooldownTime + 5000 < now) {
+		client->m_castCooldownTime = now, client->m_castCount = 0;
+	} else if (client->m_castCount++ >= 3) {
+		client->sendTextMessage(TextMessage(MESSAGE_STATUS, fmt::format("Please wait a {} seconds to send another message.", ((client->m_castCooldownTime + 5000 - now) / 1000) + 1)));
```
Typo:
```suggestion
		client->sendTextMessage(TextMessage(MESSAGE_STATUS, fmt::format("Please wait {} seconds to send another message.", ((client->m_castCooldownTime + 5000 - now) / 1000) + 1)));
```

jhogberg commented on src/creatures/players/cast/cast_viewer.cpp:
```diff
@@ -0,0 +1,1533 @@
+							client->sendTextMessage(TextMessage(MESSAGE_FAILURE, "There is already someone with that name."));
+						}
+					} else {
+						client->sendTextMessage(TextMessage(MESSAGE_FAILURE, "It is not possible very long name."));
```
```suggestion
						client->sendTextMessage(TextMessage(MESSAGE_FAILURE, "Name must be shorter than 18 letters."));
```

("18 characters" would be more correct, but I figured "letter" is easier for the target audience to understand)

jhogberg commented on src/creatures/players/cast/cast_viewer.cpp:
```diff
@@ -0,0 +1,1533 @@
+						client->sendTextMessage(TextMessage(MESSAGE_FAILURE, "It is not possible very long name."));
+					}
+				} else {
+					client->sendTextMessage(TextMessage(MESSAGE_FAILURE, "It is not possible very small name."));
```
```suggestion
					client->sendTextMessage(TextMessage(MESSAGE_FAILURE, "Name must be longer than 2 letters."));
```

jhogberg commented on src/creatures/players/cast/cast_viewer.cpp:
```diff
@@ -0,0 +1,1533 @@
+	}
+
+	auto viewerId = m_viewers.size() + 1;
+	std::string guestString = fmt::format("Guest-{}", viewerId);
```
This may lead to several viewers with the same name (`Guest-1` joins, `Guest-2` joins, `Guest-1` leaves... Next person to join gets the name `Guest-2`). Perhaps this should be a per-stream sequence number instead?

dudantas replied:
I considered this approach, but I have concerns about how it would impact the viewer 'record'. Simply incrementing the ID continuously would leave gaps in the sequence when viewers leave, which feels inefficient and messy. Using a per-stream sequence solves the issue of duplicate names, but it doesn't address the problem of maintaining a clean, continuous ID system for current viewers.

jhogberg replied:
Why is it inefficient, and why do we care about a continuous range? What am I missing? :D

Edit: if it's about 9d44a9231a5c2f5b89a6307005715776216dc647, that uses `m_viewers.size()` right? Shouldn't that be unaffected by the names?

dudantas replied:
> Why is it inefficient, and why do we care about a continuous range? What am I missing? :D
> 
> Edit: if it's about [9d44a92](https://github.com/opentibiabr/canary/commit/9d44a9231a5c2f5b89a6307005715776216dc647), that uses `m_viewers.size()` right? Shouldn't that be unaffected by the names?

True, I didn't remember that. Thank you.

#Edit
Changed: https://github.com/opentibiabr/canary/pull/2653/commits/19286fc8249ad3e33adc279baba6743454256786

jhogberg commented on src/creatures/players/cast/cast_viewer.cpp:
```diff
@@ -0,0 +1,1533 @@
+		return;
+	}
+
+	const int64_t &now = OTSYS_TIME();
```
Nitpick: `OTSYS_TIME` lands in `std::chrono::system_clock` which is not monotonic, leading to funny behavior during daylight savings changes or other adjustments to time.

Other uses appear to have the same problem. Is there any reason it doesn't use `std::chrono::steady_clock`?

dudantas replied:
You're right. Using std::chrono::steady_clock makes more sense here since it avoids issues related to daylight savings and system time adjustments. I’ll make the change to ensure more consistent behavior across the code.

#Edit
Fixed here: e8f5e4d89

jhogberg replied:
Going through the current uses of `OTSYS_TIME`, they all seem to be used for relative time ("schedule X in 2 seconds") rather than absolute time ("schedule Y at 14:00"). Perhaps it's cleaner to use `std::chrono::steady_clock` in `OTSYS_TIME`?

jhogberg reviewed (commented):
Tibiacast author here, I was having a trip down memory lane when I stumbled on this PR. I figured I'd leave a review. :slightly_smiling_face:

majestyotbr commented:
Update your pull request and use the pull request template.

dudantas committed:
Merge branch 'main' into dudantas/feat-casting-system

dudantas committed:
fix: client owner

dudantas committed:
fix: some cast bugs

github-actions[bot] committed:
fix: review from jhogberg, implement async viewers calls

dudantas committed:
enhance: add OTSYS_STEADY_TIME function

github-actions[bot] committed:
fix: message record

dudantas committed:
fix: save cast live record

dudantas committed:
fix: revert

dudantas committed:
fix: cast logout and others

dudantas committed:
fix: rename from "CastViewer" to "Livestream"

dudantas committed:
fix: send store empty

The store, by default, always opens when you press the button, it is a customer action, this way we prevent it from opening with information

dudantas commented:
> Something I notice:
> 
> Problem / Print
> When the streamer opens the gamestore window, the spectators see it too.
> Problem / Print
> Spectator opening menus like cyclopedia, bestiary, highscore, social, compendium, forge, etc.. (but, not load the info)
> Problem / Print
> Is it possible to change the Tibia title, adding Tibia Cast suffix? like `Tibia - Playername (cast)`

The gamestore is something from the cip client, when you press the store button, it always opens, I solved this by sending the "clean" data to the viewer, the window will still open, but everything will be empty (2124c6cbe). The same for cyclopedia.

About this: ```Is it possible to change the Tibia title, adding Tibia Cast suffix? like Tibia - Playername (cast)```
I don't know if it works, it needs to be investigated.

dudantas commented:
> its working cast <- only when you want to loggin to play normally on tibia 11, doesn't work

It's ok here: 
![image](https://github.com/user-attachments/assets/f9264d79-1538-4b88-a64d-e3fbbf9430f8)
![image](https://github.com/user-attachments/assets/e821f0b9-1300-4133-bbcb-d8011977c916)

dudantas commented:
> I have two active pull requests for Soul War and Cast System, and I'm not sure if this is causing interference. Specifically, with the file cast_system.lua, the file prevents players from gaining XP, and elemental weapons do not deal elemental damage.
> 
> Other problems: The player who is viewing the cast can open the following tabs: Skills, Battle List, Party List, VIP, Unjustified Points, Prey, Exaltation Forge, and Spell List, showing some information with the player's balance and others.

I couldn't open Prey and Exaltation, just like the wheel, some others can be opened because they are "client side" features, what I can do is prevent them from seeing sensitive information, like money and others, but preventing the window from opening is not possible for all of them.

Experience bug fixed here: ae07440df

dudantas committed:
fix: remove update resource balances to cast viewer

dudantas committed:
Merge branch 'main' into dudantas/feat-casting-system

dudantas committed:
fix: eventcallback name

dudantas committed:
fix: multiplier to 1.0 and experience gain fix

dudantas committed:
fix: luaConfigManagerGetFloat correct pushnumber

github-actions[bot] committed:
Code format - (Clang-format)

dudantas committed:
fix: warning for eventcallback return

github-actions[bot] committed:
Code format - (Clang-format)

dudantas committed:
Merge branch 'main' into dudantas/feat-casting-system

dudantas committed:
fix: viewer id uses per-stream sequence number

dudantas committed:
fix: change from GenericParallel to Serial

dudantas committed:
revert: remove wrong documentation

dudantas committed:
fix: restricts the viewer try to move

dudantas committed:
Merge branch 'main' into dudantas/feat-casting-system

dudantas committed:
feat: move folders, change from "cast" to "livestream" and features.hpp to disable

dudantas committed:
fix: qodana reviews

dudantas committed:
fix: qodana reviews

dudantas committed:
fix: qodana

dudantas committed:
fix: experience gain with stream on

dudantas committed:
fix: player vip check owner

phacUFPE committed:
Merge branch 'main' into dudantas/feat-casting-system

phacUFPE committed:
fix: add missing files to vcxproj visual studio

dudantas committed:
fix: indentation

dudantas committed:
Merge branch 'main' into dudantas/feat-casting-system

TheLionEscanor commented:
![Captura](https://github.com/user-attachments/assets/0d987158-fdb2-4284-9194-fbcfd9f2cedf)
This only happens with client 13.40 after a short moment of being a spectator, in OTClient it is stable

dudantas commented:
> ![Captura](https://private-user-images.githubusercontent.com/75776902/370492794-0d987158-fdb2-4284-9194-fbcfd9f2cedf.JPG?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjcyMTg5NjUsIm5iZiI6MTcyNzIxODY2NSwicGF0aCI6Ii83NTc3NjkwMi8zNzA0OTI3OTQtMGQ5ODcxNTgtZmRiMi00Mjg0LTkxOTQtZmJjZmQ5ZjJjZWRmLkpQRz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI0VDIyNTc0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZjMTFjNGIwMWNlMmUwY2VkZDcxZmJiN2ZjMjJlMWU3NmM1MzBlNWU5ZTkxZWRiMGE0YzY5ZjUyNjI5MzZkNjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.GVPcSjVyTPxmcGfMldOQcxR-w_dM5-xsKaQCMHrKSZA) This only happens with client 13.40 after a short moment of being a spectator, in OTClient it is stable

Please provide more details, because I haven't been able to find any problems yet and the debug log doesn't have the debug hash.

What were you doing at the time? Where were you? Did you perform any action? etc.

dudantas committed:
fix: livestream caster can talk to channel and show level

dudantas committed:
fix: viewer send message only in livestream channel

TheLionEscanor commented:
> > ![Captura](https://private-user-images.githubusercontent.com/75776902/370492794-0d987158-fdb2-4284-9194-fbcfd9f2cedf.JPG?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjcyMTg5NjUsIm5iZiI6MTcyNzIxODY2NSwicGF0aCI6Ii83NTc3NjkwMi8zNzA0OTI3OTQtMGQ5ODcxNTgtZmRiMi00Mjg0LTkxOTQtZmJjZmQ5ZjJjZWRmLkpQRz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI0VDIyNTc0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZjMTFjNGIwMWNlMmUwY2VkZDcxZmJiN2ZjMjJlMWU3NmM1MzBlNWU5ZTkxZWRiMGE0YzY5ZjUyNjI5MzZkNjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.GVPcSjVyTPxmcGfMldOQcxR-w_dM5-xsKaQCMHrKSZA) This only happens with client 13.40 after a short moment of being a spectator, in OTClient it is stable
> 
> Please provide more details, because I haven't been able to find any problems yet and the debug log doesn't have the debug hash.
> 
> What were you doing at the time? Where were you? Did you perform any action? etc.

It was a packet problem, for now it is stable, but now i cant enter to the livestream in OTClient, is just waiting for loading the livestreams table throwing this:
ERROR: Unknown data type 45
ERROR: Unknown data type 71
ERROR: protected lua call failed: /modules/corelib/inputmessage.lua:31: table index is nil
stack traceback:
	[C]: in function '__newindex'
	/modules/corelib/inputmessage.lua:31: in function 'getTable'
	/modules/gamelib/protocollogin.lua:286: in function 'parseExtendedCharacterList'
	/modules/gamelib/protocollogin.lua:183: in function </modules/gamelib/protocollogin.lua:163>

dudantas committed:
Merge branch 'main' into dudantas/feat-casting-system

dudantas committed:
Merge branch 'main' into dudantas/feat-casting-system

dudantas committed:
fix: compilation

dudantas committed:
fix: some sonar checks

dudantas committed:
fix: sonar checks

OpenTibiaMX commented:
Does anyone have the login.php for Gesior AAC Global that they can share with me?

Mattz0n commented:
Is this cast system working on Canary 1340 ? or is it outdated as it stays?

Feliipe93 commented:
is this working or somebody has? this is being sold

dudantas commented:
> is this working or somebody has? this is being sold

Haven't you tested it? It's in the pull request description that it needs to be tested to be merged. I haven't seen almost anyone's review here, instead of buying it, shouldn't they test it here and give feedback?
